demos/security_barrier_camera_demo/main.cpp

```diff
@@ -333,27 +335,39 @@ void Drawer::process() {
         lastShownframeId++;
         cv::Mat mat = firstGridIt->second.getMat();
 
-        float opacity = 0.6f;
-        fillROIColor(mat, cv::Rect(5, 5, 700, 115), cv::Scalar(255, 0, 0), opacity);
-
-        std::ostringstream out;
-        out << "Mean overall time per all inputs: " << std::fixed << std::setprecision(2) << std::setw(6);
-        const auto t1 = std::chrono::steady_clock::now();
-        uint64_t frameCounter = context.frameCounter;
-        const ms meanOverallTimePerAllInputs = std::chrono::duration_cast<ms>((t1 - context.t0)
-                                               * context.readersContext.inputChannels.size()) / frameCounter;
-        out << meanOverallTimePerAllInputs.count();
-        out << "ms /" << std::setw(6) << std::chrono::seconds(1) / meanOverallTimePerAllInputs << "FPS";
-
-        cv::putText(mat, out.str(), cv::Point2f(10, 35), cv::FONT_HERSHEY_TRIPLEX, 0.7, cv::Scalar{255, 255, 255});
-        cv::putText(mat, "Detection InferRequests usage", cv::Point2f(10, 70), cv::FONT_HERSHEY_TRIPLEX, 0.7, cv::Scalar{255, 255, 255});
-        cv::Rect usage(15, 90, 400, 20);
+        constexpr float OPACITY = 0.6f;
+        fillROIColor(mat, cv::Rect(5, 5, 390, 115), cv::Scalar(255, 0, 0), OPACITY);
+        cv::putText(mat, "Detection InferRequests usage", cv::Point2f(15, 70), cv::FONT_HERSHEY_TRIPLEX, 0.7, cv::Scalar{255, 255, 255});
+        cv::Rect usage(15, 90, 370, 20);
         cv::rectangle(mat, usage, {0, 255, 0}, 2);
         uint64_t nireq = context.nireq;
+        uint64_t frameCounter = context.frameCounter;
         usage.width = static_cast<int>(usage.width * static_cast<float>(frameCounter * nireq - context.freeDetectionInfersCount) / (frameCounter * nireq));
         cv::rectangle(mat, usage, {0, 255, 0}, cv::FILLED);
 
-        cv::imshow("Detection results", firstGridIt->second.getMat());
+        static std::ostringstream outThroughput;
+        static unsigned fps = 0;
+        static bool runningFirstSecond = true;
+        static std::chrono::steady_clock::time_point localT0 = context.t0;
+        fps++;
+        const std::chrono::steady_clock::time_point localT1 = std::chrono::steady_clock::now();
+        const Sec timeDuration = localT1 - localT0;
+        if (std::chrono::seconds{1} <= timeDuration) {
+            outThroughput.str("");
+            outThroughput << std::fixed << std::setprecision(1)
+                << static_cast<float>(fps) / timeDuration.count() << "FPS";
+            fps = 0;
+            runningFirstSecond = false;
+            localT0 = localT1;
+        } else if (runningFirstSecond) {
+            // the pipeline isn't running for long enough, print what it has for now
+            outThroughput.str("");
+            outThroughput << std::fixed << std::setprecision(1)
+                << static_cast<float>(fps) / timeDuration.count() << "FPS";
+        }
```

Comment:
You've read this right.

We can't put empty string because the demo supports a single image input. This means it will show nothing for FPS for the whole pipeline.

As for running average, the demo runs slower for first frames because for better performance it should feed all steps. And it will be hard to say when it is OK to begin relying on the reported number. I also don't like this approach because it can be affected by some user activities. For example a user generates a load for a system for a short time, but running average will be affected and again it is hard to tell when it will be eliminated.

> or on every N frames

This is what I am doing but for every N (1) seconds instead of frames.

Reply:
> We can't put empty string because the demo supports a single image input. This means it will show nothing for FPS for the whole pipeline.

Well... does a metric like FPS even make sense for a single image? You yourself said that the pipeline runs slower at first, so showing an FPS number would be misleading.

> As for running average, the demo runs slower for first frames because for better performance it should feed all steps. And it will be hard to say when it is OK to begin relying on the reported number. I also don't like this approach because it can be affected by some user activities. For example a user generates a load for a system for a short time, but running average will be affected and again it is hard to tell when it will be eliminated.

Aren't the same problems present with any method of framerate calculation? You're solving them by ignoring all frames that are older than 1 second; whereas a running average solves them by ignoring all frames except the N most recent ones.

Reply:
I messed up running average with alpha blending.

I am going to implement [running average](https://en.wikipedia.org/wiki/Moving_average) with two parameters: frames before FPS metric update (`-fbu`) (number of frames) which will represent how frequently FPS metric will be updated in terms of `imshow()` and FPS metric queue size (`-fqs`) (number of frames) which is the number of last frames (`imshow`s) to take into account.

@IRDonch, @vmatveic, what do you think?

Reply:
Do we really need these to be parameterizable? IMO, you should pick values that produce sensible results and stick with those.

Reply:
This will allow to keep the demo backward compatible. Setting `-fqs` to 0 (unlimited) and `-fbu` to 1 will do the trick. On the other hand setting `-fqs` and `-fbu` to N will behave like the current implementation in term of frames instead of seconds{1}.

Reply:
> This will allow to keep the demo backward compatible.

Well... do we actually need to do that? The _current_ version doesn't preserve backwards compatibility, so evidently you're fine with that. :-) I don't think the original behavior is all that useful, so I don't see why we'd need to preserve it.

Reply:
It only removes millisecond metric. This demo is used in `openvino/deployment_tools/demo/demo_security_barrier_camera.sh` with an image. I am not sure that we can remove this behavior.

Reply:
>  I am not sure that we can remove this behavior.

I don't know which behavior you mean, but when the demo is run on a single image, wouldn't a running average (with any parameters) be compatible with existing behavior, as long as you make sure it updates on the first frame?

Reply:
Would it be an acceptable solution to keep the current behavior but handle the first frame differently? It will update the metric on the first frame and every second after that.

Reply:
Sure, that seems reasonable.

Reply:
Done